xmppserver/src/main/java/org/jivesoftware/openfire/server/ServerDialback.java

```diff
@@ -473,35 +491,31 @@ protected void dialbackError(String from, String to, PacketError err) {
      *
      * @param doc the request for validating the new domain.
      * @param streamID the stream id generated by this server for the Originating Server.
-     * @return true if the requested domain is valid.
+     * @throws StreamErrorException when validation did not succeed.
      */
-    public boolean validateRemoteDomain(Element doc, StreamID streamID) {
+    public void validateRemoteDomain(Element doc, StreamID streamID) throws StreamErrorException, ServerDialbackErrorException, ServerDialbackKeyInvalidException
+    {
         StringBuilder sb;
         String recipient = doc.attributeValue("to");
         String remoteDomain = doc.attributeValue("from");
 
         final Logger log = LoggerFactory.getLogger( Log.getName() + "[Acting as Receiving Server: Validate domain: " + recipient + "(id " + streamID + ") for OS: " + remoteDomain + "]" );
 
         log.debug( "Validating domain...");
-        if (connection.getTlsPolicy() == Connection.TLSPolicy.required &&
-                !connection.isEncrypted()) {
-            connection.deliverRawText(new StreamError(StreamError.Condition.policy_violation).toXML());
-            // Close the underlying connection
-            connection.close();
-            return false;
+
+        if (!ServerDialback.isEnabled() && !ServerDialback.isEnabledForSelfSigned()) {
+            throw new StreamErrorException(new StreamError(StreamError.Condition.policy_violation, "Dialback has been disabled by configuration."));
```

Comment:
Should the message be set on the StreamErrorException as well as the StreamError?
Or when this gets logged (e.g. on on L428) will the logger reach into StreamError to get its internal message?

(Of course, we could update the StreamErrorException constructor to set the Exception message to the Error's message if it is provided and the exception's message isn't, but it's getting close to fallable logic inside error handling, which is dangerous)

Reply:
I think it's a good idea to provide as much context as reasonabily possible. That will help diagnosing problems. I have added some changes to the constructor. I believe they're fairly safe, but givin your last remark, it might be wise to review that carefully.

L428 is inside a method that doesn't seem to be used (I'm contemplating to remove it completely, but I wonder if maybe it _should_ be used somewhere, which is why I've not done that yet). In that method, the error isn't being transmitted to the client, nor is the connection closed. I've added that, too.

Reply:
Ace! I've added a comment on the constructor change separately, further down the PR :)